tuf/api/metadata.py

```diff
@@ -1435,6 +1444,149 @@ def is_delegated_path(self, target_filepath: str) -> bool:
+        threshold = role_dict.pop("threshold")
+        bit_length = role_dict.pop("bit_length")
+        name_prefix = role_dict.pop("name_prefix")
+        # All fields left in the role_dict are unrecognized.
+        return cls(keyids, threshold, bit_length, name_prefix, role_dict)
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Returns the dict representation of self."""
+        base_role_dict = super().to_dict()
+        return {
+            "bit_length": self.bit_length,
+            "name_prefix": self.name_prefix,
+            **base_role_dict,
+        }
+
+    def get_bin_name(self, bin_number: int = 0) -> str:
+        """Return the name of bin with number ``bin_number``.
+
+        The ``bin_number`` default is 0 as that is used to uniquely identify
+        delegated roles containing succinct_roles from the rest.
+
+        Args:
+            bin_number: Number of the bin that we want the name for.
+        """
+        # Add zero padding if necessary and cast to hex the suffix.
+        suffix = f"{bin_number:0{self.suffix_len}x}"
+        return f"{self.name_prefix}-{suffix}"
+
+    def find_bin(self, target_filepath: str) -> str:
+        """Calculates the name of the bin responsible for ``target_filepath``.
+
+        Args:
+            target_filepath: URL path to a target file, relative to a base
+                targets URL.
+        """
+        hasher = sslib_hash.digest(algorithm="sha256")
+        hasher.update(target_filepath.encode("utf-8"))
+
+        # We can't ever need more than 4 bytes (32 bits).
+        hash_bytes = hasher.digest()[:4]
+        # Right shift hash bytes, so that we only have the leftmost
+        # bit_length bits that we care about.
+        shift_value = 32 - self.bit_length
+        bin_number = int.from_bytes(hash_bytes, byteorder="big") >> shift_value
+
+        return self.get_bin_name(bin_number)
+
+    def get_all_bin_names(self) -> Iterator[str]:
+        """Yield the names of all different bins one by one."""
+        for i in range(0, self.number_of_bins):
+            yield self.get_bin_name(i)
+
+    def is_bin(self, role_name: str) -> bool:
+        """Determines whether the given ``role_name`` is in one of
+        the bins that ``SuccinctRoles`` is trusted to provide.
+
+        Args:
+            role_name: The name of the delegated role to check against.
+        """
+        desired_prefix = self.name_prefix + "-"
+
+        if role_name.startswith(desired_prefix):
+            suffix = role_name[len(desired_prefix) :]
+            try:
+                # bin name suffixes are in hex format.
+                num = int(suffix, 16)
+                if (
+                    0 <= num < self.number_of_bins
+                    and len(suffix) == self.suffix_len
+                ):
+                    return True
+            except ValueError:
+                # suffix is not a number
+                return False
+
+        return False
```

Comment:
this would likely improve from early returns, something like this:

```suggestion
        if not role_name.startswith(desired_prefix):
            return False

        suffix = role_name[len(desired_prefix) :]
        if len(suffix) != self.suffix_len:
            return False

        try:
            # make sure suffix is hex value
            num = int(suffix, 16)
        except ValueError:
            return False

        return num >= 0 and num < self.number_of_bins
```

Reply:
Agree and it looks more readable.
I heavily focused on what is necessary for the `role_name` to be a bin and ended up with long checks.